File: core/src/main/java/io/github/fourlastor/game/coordinates/HexCoordinates.java
package io.github.fourlastor.game.coordinates;

import com.badlogic.gdx.math.Vector2;

public class HexCoordinates {

    private final int width;
    private final int height;
    private final int hexSideLength;
    private final float horizontalSpacing;
    private final float staggerSpacing;

    public HexCoordinates(int tileWidth, int tileHeight, int hexSideLength) {
        this.width = tileWidth;
        this.height = tileHeight - 1;
        this.hexSideLength = hexSideLength;
        this.horizontalSpacing = (tileWidth - hexSideLength) / 2f + hexSideLength - 1;
        this.staggerSpacing = height / 2f;
    }

    public Vector2 toWorldAtOrigin(int x, int y, Vector2 out) {
        float stagger = x % 2 == 0 ? 0f : staggerSpacing;
        return out.set(x * horizontalSpacing, stagger + y * height);
    }

    public Vector2 toWorldAtCenter(Hex hex, Vector2 out) {
        return toWorldAtCenter(hex.offset.x, hex.offset.y, out);
    }

    public Vector2 toWorldAtCenter(int x, int y, Vector2 out) {
        return toWorldAtOrigin(x, y, out).add(width / 2f, height / 2f);
    }
}

File: core/src/main/java/io/github/fourlastor/game/coordinates/HexCoordinates.kt
package io.github.fourlastor.game.coordinates

import com.badlogic.gdx.math.Vector2

class HexCoordinates(private val width: Int, tileHeight: Int, hexSideLength: Int) {
    private val height: Int
    private val horizontalSpacing: Float
    private val staggerSpacing: Float

    init {
        height = tileHeight - 1
        horizontalSpacing = (width - hexSideLength) / 2f + hexSideLength - 1
        staggerSpacing = height / 2f
    }

    fun toWorldAtOrigin(x: Int, y: Int, out: Vector2): Vector2 {
        val stagger = if (x % 2 == 0) 0f else staggerSpacing
        return out.set(x * horizontalSpacing, stagger + y * height)
    }

    fun toWorldAtCenter(hex: Hex): Vector2 {
        return toWorldAtCenter(hex.offset.x, hex.offset.y, Vector2())
    }

    fun toWorldAtCenter(hex: Hex, out: Vector2): Vector2 {
        return toWorldAtCenter(hex.offset.x, hex.offset.y, out)
    }

    fun toWorldAtCenter(x: Int, y: Int): Vector2 {
        return toWorldAtCenter(x, y, Vector2())
    }

    fun toWorldAtCenter(x: Int, y: Int, out: Vector2): Vector2 {
        return toWorldAtOrigin(x, y, out).add(width / 2f, height / 2f)
    }
}
